### DISCUSSION TITLE
Kivy-ios 'hello world' fails to build to simulator on new M2 Mac mini with MacOS 14.4.1 and XCode 15.3

### DISCUSSION THREAD

**Discussion by coryholl-kivy** [kivy/kivy-github-discussions#35] [Q&A]
I created a 'hello world' app to test kivy-ios on a new mac mini and have been unable to get it to build in XCode.  I fear I may have a "too new" problem where kivy-ios is not compatible with this new dev environment.

I followed the instructions on https://github.com/kivy/kivy-ios and have reached the "click Play" in XCode to start the simulator and get a unresolved symbols build failure.

The hello world program is **very** simple.
main.py:
```
import kivy
from kivy.app import App

class MainApp(App):
    pass

if __name__ == '__main__':
    app = MainApp()
    app.run()
```

main.kv:
```
Label:
    text: 'Hello World!'
```

When I run XCode as so:
```$ open helloworld-ios/helloworld.xcodeproj/```

I get this error:
![error screenshot1](https://github.com/kivy/kivy-github-discussions/assets/168998651/7f1656f1-fb2f-4e3e-860b-c43b7f1472a2)

```Showing Recent Issues
Ld /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Products/Debug-iphonesimulator/helloworld.app/helloworld normal (in target 'helloworld' from project 'helloworld')
    cd /Users/dalek/toolchain/helloworld-ios
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -Xlinker -reproducible -target arm64-apple-ios9.0-simulator -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator17.4.sdk -O0 -L/Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/EagerLinkingTBDs/Debug-iphonesimulator -L/Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Products/Debug-iphonesimulator -L/Users/dalek/toolchain/dist/lib -L/Users/dalek/toolchain/dist/lib/iphoneos -L/Users/dalek/toolchain/dist/lib/iphonesimulator -L/Users/dalek/toolchain/dist/lib -L/Users/dalek/toolchain/dist/lib/iphoneos -L/Users/dalek/toolchain/dist/lib/iphonesimulator -F/Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/EagerLinkingTBDs/Debug-iphonesimulator -F/Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Products/Debug-iphonesimulator -F/Users/dalek/toolchain/dist/frameworks -F/Users/dalek/toolchain/helloworld-ios/../dist/xcframework -filelist /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/Objects-normal/arm64/helloworld.LinkFileList -Xlinker -object_path_lto -Xlinker /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/Objects-normal/arm64/helloworld_lto.o -Xlinker -no_deduplicate -Xlinker -objc_abi_version -Xlinker 2 -fobjc-link-runtime -v -Xlinker -sectcreate -Xlinker __TEXT -Xlinker __entitlements -Xlinker /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/helloworld.app-Simulated.xcent -Xlinker -sectcreate -Xlinker __TEXT -Xlinker __ents_der -Xlinker /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/helloworld.app-Simulated.xcent.der -framework Photos -framework Metal -framework AVFoundation -framework QuartzCore -framework CoreMedia -framework WebKit -framework Accelerate -framework AudioToolbox -framework GameController -framework OpenGLES -framework MobileCoreServices -framework CoreVideo -framework ImageIO -framework MessageUI -framework CoreHaptics -framework CoreMotion -framework UIKit -framework CoreGraphics -lc++ -lbz2 -lsqlite3 -lz -Xlinker -no_adhoc_codesign -Xlinker -dependency_info -Xlinker /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/Objects-normal/arm64/helloworld_dependency_info.dat -o /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Products/Debug-iphonesimulator/helloworld.app/helloworld

Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-ios9.0-simulator
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" -demangle -lto_library /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib -no_deduplicate -dynamic -arch arm64 -platform_version ios-simulator 14.0.0 17.4 -syslibroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator17.4.sdk -O0 -o /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Products/Debug-iphonesimulator/helloworld.app/helloworld -L/Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/EagerLinkingTBDs/Debug-iphonesimulator -L/Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Products/Debug-iphonesimulator -L/Users/dalek/toolchain/dist/lib -L/Users/dalek/toolchain/dist/lib/iphoneos -L/Users/dalek/toolchain/dist/lib/iphonesimulator -L/Users/dalek/toolchain/dist/lib -L/Users/dalek/toolchain/dist/lib/iphoneos -L/Users/dalek/toolchain/dist/lib/iphonesimulator -reproducible -filelist /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/Objects-normal/arm64/helloworld.LinkFileList -object_path_lto /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/Objects-normal/arm64/helloworld_lto.o -no_deduplicate -objc_abi_version 2 -sectcreate __TEXT __entitlements /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/helloworld.app-Simulated.xcent -sectcreate __TEXT __ents_der /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/helloworld.app-Simulated.xcent.der -framework Photos -framework Metal -framework AVFoundation -framework QuartzCore -framework CoreMedia -framework WebKit -framework Accelerate -framework AudioToolbox -framework GameController -framework OpenGLES -framework MobileCoreServices -framework CoreVideo -framework ImageIO -framework MessageUI -framework CoreHaptics -framework CoreMotion -framework UIKit -framework CoreGraphics -lc++ -lbz2 -lsqlite3 -lz -no_adhoc_codesign -dependency_info /Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/helloworld.build/Debug-iphonesimulator/helloworld.build/Objects-normal/arm64/helloworld_dependency_info.dat -framework Foundation -lobjc -lSystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/15.0.0/lib/darwin/libclang_rt.iossim.a -F/Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Intermediates.noindex/EagerLinkingTBDs/Debug-iphonesimulator -F/Users/dalek/Library/Developer/Xcode/DerivedData/helloworld-ehvsrgsndqapquboabscwhyjfukg/Build/Products/Debug-iphonesimulator -F/Users/dalek/toolchain/dist/frameworks -F/Users/dalek/toolchain/helloworld-ios/../dist/xcframework
Undefined symbols for architecture arm64:
  "_PyEval_InitThreads", referenced from:
      _SDL_main in main.o
  "_PyMem_RawMalloc", referenced from:
      _SDL_main in main.o
  "_PyRun_SimpleFileExFlags", referenced from:
      _SDL_main in main.o
  "_PyRun_SimpleStringFlags", referenced from:
      _load_custom_builtin_importer in main.o
  "_PySys_SetArgv", referenced from:
      _SDL_main in main.o
  "_Py_DecodeLocale", referenced from:
      _SDL_main in main.o
  "_Py_Finalize", referenced from:
      _SDL_main in main.o
  "_Py_Initialize", referenced from:
      _SDL_main in main.o
  "_main", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Undefined symbol: _PyEval_InitThreads

Undefined symbol: _PyMem_RawMalloc

Undefined symbol: _PyRun_SimpleFileExFlags

Undefined symbol: _PyRun_SimpleStringFlags

Undefined symbol: _PySys_SetArgv

Undefined symbol: _Py_DecodeLocale

Undefined symbol: _Py_Finalize

Undefined symbol: _Py_Initialize

Undefined symbol: _main

Linker command failed with exit code 1 (use -v to see invocation)
```

What I know is that main.o is being linked to libSDL2.a to resolve they symbol for SDL_main but the linker is failing to link libpython3.11.a into the build to resolve the symbols in libSDL2.a.  I have been unable to figure out how to get XCode to perform the correct link.

I am thinking this is a regression issue, otherwise I would expect other people to be reporting this issue.  But I cannot confirm this as I do not have access to older MacOS environments.

Anyone have any suggestions?  I would be more than happy to assist in helping get kivy-ios working for MacOS 14/XCode 15.3.

**Comment by spyke555** [comment]
Well, I am a little ahead of you on Xcode 15.4 and macOS 14.5. I am getting a little further in my build process but fail after about 29 seconds with a "rsync warning: some files vanished before they could be transferred (code 24)" error code. Platform build and xcode project creation steps seem to have occured correctly. Seems to me this was working last week, but I've reset my macOS and re-installed fresh since to try and isolate the issue.

Note that I am developing on a Mac mini M1 device.

Good to know I am not alone...

**Comment by cederom** [comment]
yeah XCode 15 has different build mechanism from previous releases. I am unable to produce IPA that would be able to publish in the store. Emulator build works but then application crashes (see issues it is a known problem) still the same application runs on a physical device.

Not being able to store publish is a biggest PITA :-(

**Comment by coryholl-kivy** [comment]
I have confirmed that upgrading to XCode 15.4 does not change the build issue.  My hello world program fails to build with the same linker errors.

I will try it again when either XCode or kivy-ios gets another update.

**Comment by cederom** [comment]
Probably some build dependencies / packages / tools / recipes are missing (target python recipe?).

Try using buildozer:
* `buildozer ios update` - this will pull add build dependencies.
* `buildozer ios debug deploy` - this will create the xcodeproject template (in `.buildozer`).
* `buildozer ios xocde` - this will launch the xcode with your project, there you can setup stuff, select target device, build and run.

Notes:
1. There is a known problem when running on simulator application crashes. Try using real iPhone over USB cable that works for me.
2. I am still unable to produce publishable IPA - there are some build artifacts that needs to be removed by hand, also IOS_TARGET is still 9 while it should be 12 and that requires some icons modification.

**Comment by coryholl-kivy** [comment]
Buildozer on Apple is more broken than kivy-ios I am afraid.  You are the first person who suggested I use it as most people recommend avoiding it.  I have been using Buildozer on Android for the last 8 months successful, the iOS logic needs some TLC however.

I tried again after upgrading to OSX 14.7 with the same results.  I didn't expect any changes.  I am hesitant to upgrade to OSX 15 and XCode 16 as I am afraid the tools will just get further and further away from a compatible software combination.

I will try again in a month or two and see if the lie of the land has changed in the iOS Kivy camp.

**Comment by Einehordebedrohenie** [comment]
Hello! I am also unable to build my app on Xcode. I get "Command PhaseScriptExecution failed with a nonzero exit code". Xcode 15.2, MacMini M1. Any Ideas? Is the iOS-Kivy-thing still supported and updated? I mean am I trying something that should normally work? Thanks!

**Reply by coryholl-kivy** [reply]
I saw a patch being applied to Kivy-IOS for Xcode 16 a few weeks ago to the trunk.  I have to wonder if the mainline development line might be working with higher versions of Xcode on the M architecture.

I am watching to see if the next stable Kivy-IOS release works.  I am hoping it contains patches for the newer tools and hardware architectures.

**Comment by akshayaurora** [comment]
This is not an issue tracker. Please use this forum only for discussions related to the framework , not for issues related to particular repository

kivy-ios has been working with Xcode, including being able to publish to iOS App Store.
Because this discussion/issue happened in the wrong place, there was no one to help out/respond to this. 

Please Use latest kivy-ios from master, Use appropriate channels for help/issues either issue list on kivy-ios or buildozer or the appropriate Kivy-users forum or the discord # support channels for getting help related to such issues.
